src/java.base/share/classes/java/util/HashMap.java

```diff
@@ -746,7 +746,7 @@ else if (e instanceof TreeNode)
                         }
                         if (hiTail != null) {
                             hiTail.next = null;
-                            newTab[j + oldCap] = hiHead;
+                            newTab[j | oldCap] = hiHead;
```

Comment:
`oldCap` can be values like 13, which is not a power of 2, and the bitwise or here is not equivalent to an addition.

Reply:
else if ((newCap = oldCap << 1) < MAXIMUM_CAPACITY &&
                     oldCap >= DEFAULT_INITIAL_CAPACITY)
                newThr = oldThr << 1; // double threshold
                
                static final int MAXIMUM_CAPACITY = 1 << 30 = 1073741824
                static final int DEFAULT_INITIAL_CAPACITY = 1 << 4 = 16
                
The above code snippet is for initialization and array resizing, starting from DEFAULT_INITIAL_CAPACITY. Each time, oldCap is left-shifted by one position, ensuring that it always remains a power of two. Consequently, bitwise OR operations can be used instead of addition.

Reply:
You are quoting the wrong code but `tableSize()` generates a power-of-2 size and the javadoc says "move with a power of 2 offset" so it's fine.

Reply:
Thanks to the mailing list guy who reveals the responose from @Yumfeil is made with ChatGPT. This is a bad-faith patch, it's just lucky that it's simple and has no real damage.

Reply:
I personally find it more readable to use | than +, since this is thinking in binary rather than decimal. As for that reply, my English is not very good, so I used GLM4 for auxiliary translation, not ChatGPT

This is my reason